### ISSUE TITLE
[BUG] Dots animation stops

### ISSUE THREAD

**Issue by jzielke84** [flixlix/power-flow-card-plus#131]
**Describe the bug**
Animation of dots stops after some time and only resumes on reload of browser. Watts values in the card keep changing constantly though.

**To Reproduce**
1. Add card to dashboard
2. Watch dashboard for a couple of minutes
3. See animation stop (but values still changing)
4. Reload browser

**Expected behavior**
Continue dot animation idenfinately. 

**Desktop (please complete the following information):**
 - OS: iOS, Android, OSX... choose what you want
 - Browser: Chrome
 - Version latest

**Smartphone (please complete the following information):**
see "Desktop"

**Log
Chrome console throws the following:

```java
power-flow-card-plus.js?hacstag=6180818150096:322 Uncaught (in promise) TypeError: Failed to execute 'setCurrentTime' on 'SVGSVGElement': The provided float value is non-finite.
    at power-flow-card-plus.js?hacstag=6180818150096:322:17551
    at Array.forEach (<anonymous>)
    at ae.render (power-flow-card-plus.js?hacstag=6180818150096:322:17437)
    at ae.update (power-flow-card-plus.js?hacstag=6180818150096:23:294)
    at o.update (card-mod.js?hacstag=190927524322:5:1322)
    at ae.performUpdate (power-flow-card-plus.js?hacstag=6180818150096:11:5085)
    at ae.scheduleUpdate (power-flow-card-plus.js?hacstag=6180818150096:11:4732)
    at ae._$Ej (power-flow-card-plus.js?hacstag=6180818150096:11:4640)
(anonymous) @ power-flow-card-plus.js?hacstag=6180818150096:322
render @ power-flow-card-plus.js?hacstag=6180818150096:322
update @ power-flow-card-plus.js?hacstag=6180818150096:23
o.update @ card-mod.js?hacstag=190927524322:5
performUpdate @ power-flow-card-plus.js?hacstag=6180818150096:11
scheduleUpdate @ power-flow-card-plus.js?hacstag=6180818150096:11
_$Ej @ power-flow-card-plus.js?hacstag=6180818150096:11
await in _$Ej (async)
requestUpdate @ power-flow-card-plus.js?hacstag=6180818150096:11
set @ power-flow-card-plus.js?hacstag=6180818150096:11
value @ hui-stack-card.ts:63
_$AE @ reactive-element.ts:1380
performUpdate @ reactive-element.ts:1345
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
value @ hui-stack-card.ts:63
_$AE @ reactive-element.ts:1380
performUpdate @ reactive-element.ts:1345
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
(anonymous) @ hui-view.ts:151
value @ hui-view.ts:149
performUpdate @ reactive-element.ts:1331
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
value @ hui-root.ts:598
_$AE @ reactive-element.ts:1380
performUpdate @ reactive-element.ts:1345
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
j @ lit-html.ts:1871
_$AI @ lit-html.ts:1813
p @ lit-html.ts:1181
$ @ lit-html.ts:1528
_$AI @ lit-html.ts:1375
G @ lit-html.ts:2172
update @ lit-element.ts:165
performUpdate @ reactive-element.ts:1331
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
value @ partial-panel-resolver.ts:112
value @ hass-router-page.ts:93
performUpdate @ reactive-element.ts:1331
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
j @ lit-html.ts:1871
_$AI @ lit-html.ts:1813
p @ lit-html.ts:1181
$ @ lit-html.ts:1528
_$AI @ lit-html.ts:1375
G @ lit-html.ts:2172
update @ lit-element.ts:165
performUpdate @ reactive-element.ts:1331
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
j @ lit-html.ts:1871
_$AI @ lit-html.ts:1813
p @ lit-html.ts:1181
$ @ lit-html.ts:1528
_$AI @ lit-html.ts:1375
G @ lit-html.ts:2172
update @ lit-element.ts:165
performUpdate @ reactive-element.ts:1331
scheduleUpdate @ reactive-element.ts:1263
_$Ej @ reactive-element.ts:1235
await in _$Ej (async)
requestUpdate @ reactive-element.ts:1210
set @ reactive-element.ts:726
value @ hass-base-mixin.ts:52
_updateHass @ context-mixin.ts:95
(anonymous) @ connection-mixin.ts:190
s @ store.js:19
(anonymous) @ entities.js:74
e.subscribeMessage.type @ entities.js:76
(anonymous) @ connection.js:23
_handleMessage @ connection.js:15
```

**Please make sure all of this applies and check the checkboxes, by writing "X" inside of them.**

- [x] I understand that this is a channel for reporting bugs, not a support forum (If you need help configuring the card, go to: [Discussions](https://github.com/flixlix/power-flow-card-plus/discussions)).

- [X] I have the latest version of the card installed.

- [X] I have cleared my browser cache **after** installing, updating and configuring the card.

**Comment by flixlix** [comment]
I am not experiencing this behavior even after 15 minutes. After how much time did your dots stop flowing? I did experience this if an entity went offline

**Comment by jzielke84** [comment]
This is happening in conjunction with meter readings provided by https://github.com/klaasnicolaas/home-assistant-glow. However, the instance providing the data doesn't go offline. The number in the power-flow card even changes after the dot has stopped moving. This happens like after 5 minutes but don't nail me down on the exact minute.

**Comment by flixlix** [comment]
Your configuration is incorrect.
It should look like this:

```yaml
type: custom:power-flow-card-plus
entities:
  grid:
    entity: sensor.stromzahler_power_consumption
    color_circle: true
    color:
      production: var(--energy-solar-color)
kw_decimals: 1
watt_threshold: 999
max_flow_rate: 5
min_flow_rate: 2
clickable_entities: true
title: Stromverbrauch
```

Or, in case you have one entity for each direction 

```yaml
type: custom:power-flow-card-plus
entities:
  grid:
    entity: 
      consumption: sensor.stromzahler_power_consumption
      production: sensor.stromzahler_power_production
    color_circle: true
    color:
      production: var(--energy-solar-color)
kw_decimals: 1
watt_threshold: 999
max_flow_rate: 5
min_flow_rate: 2
clickable_entities: true
title: Stromverbrauch
```

**Comment by jzielke84** [comment]
Thanks for clarifying. I've changed it and will report back to you.

**Comment by flixlix** [comment]
I'm pretty busy ATM, but I'll test this behaviour and report with you later

**Comment by flixlix** [comment]
> What beta? Homeassistant?

The version 0.1beta of this card

> Have you tried this with a single sensor in your card? You have plenty of them in your example.

Tried here with one entity only, bug doesn't occur.

https://user-images.githubusercontent.com/61006057/235739422-95e716e0-c45f-4ab4-9b58-6381f1ca2815.mov

This is the config:

```yaml
type: custom:power-flow-card-plus
entities:
  grid:
    display_state: one_way
    entity: input_number.grid_power_consumption
w_decimals: 0
kw_decimals: 2
min_flow_rate: 0.9
max_flow_rate: 6
max_expected_flow_w: 0
watt_threshold: 1000000
clickable_entities: true
title: Power Flow Card Plus
```

I am using the beta version of this card

**Comment by jzielke84** [comment]
Tried everything I could:

- installed your beta version
- reinstalled the card completely
- copied your card config
- created a whole new dashboard with just this card
- create a card with a number helper

Every time once it's 0 the dot stops 😒

**Comment by flixlix** [comment]
Can you try this file? You Browser Console, should log this:

<img width="430" alt="Bildschirmfoto 2023-05-02 um 19 54 54" src="https://user-images.githubusercontent.com/61006057/235746082-560382f4-5a11-4293-b4a4-907c3a664b31.png">

Otherwise, you don't have the card correctly integrated. In this case, clear your browser's cache.

[power-flow-card-plus.js.zip](https://github.com/flixlix/power-flow-card-plus/files/11376230/power-flow-card-plus.js.zip)

**Comment by jzielke84** [comment]
Did that to no avail. It also shows up in my debug console:
![image](https://user-images.githubusercontent.com/13448042/235747361-f9832cd8-5685-4cc3-a936-00d54957eafc.png)

Browser cache has also been emptied.

**Comment by flixlix** [comment]
> When I add a new power-flow-card, the editor shows this

That is a known bug related to this issue: https://github.com/flixlix/power-flow-card-plus/issues/133

Basically it means there's a conflict between this card and the current version of card mod.

<hr/>

Now, in the case, of the console, it seems you still have the HACS version installed. Please remove that resource and add the `www/power-flow-card-plus.js` resource. Also, you need to unzip the file I sent you

**Comment by flixlix** [comment]
Nice, now you have the right version.
Is it still the same error? Shouldn't be, since I changed the line to this:

```js
flowSVGElement.setCurrentTime(flowSVGElement.getCurrentTime() * (newDur[flowName]));
```

Either that or the issue was not the dividing by 0

**Comment by jzielke84** [comment]
I don't see that line in the file you have provided me?!

**Comment by flixlix** [comment]
It's the same as the previous line that you provided to me:

```js
t.setCurrentTime(t.getCurrentTime() * (Nn[e] / this.previousDur[e])),
```

But the one I sent is not compiled yet, it's in typescript

**Comment by jzielke84** [comment]
So you want me to replace it manually? I'd like to but the file is totally cluttered. Many line breaks are missing. So better compile it and pass me the file for testing please.

**Comment by jzielke84** [comment]
Ah ok, well the error is still happening. Same line, same script message.

**Comment by flixlix** [comment]
> https://user-images.githubusercontent.com/13448042/235736662-ac51fc9b-356d-4510-9098-b7f7e67023fe.mp4

Can you send me a screen recording of this behavior? This link seems to be broken, I can't access it

**Comment by flixlix** [comment]
Thank you for updating the video format, now I can see it. What I noticed, is that when you set the state to 0, in the card, the grid displayed 290. This is also not expected behavior, maybe it has something to do with it

**Comment by jzielke84** [comment]
That's just because the esphome sensor from my power meter is still delivering data. The 0 causes the dot to stop and further updates of the sensor only change the number within the circle.

**Comment by jzielke84** [comment]
Using this partial configuration the dot also stops:
```yaml
entities:
  grid:
    entity:
      consumption: sensor.stromzahler_power_consumption
      production: input_number.testwatt
    color_circle: true
    color:
      production: var(--energy-solar-color)
```

Then I've tried the following configuration:

```yaml
entities:
  grid:
    entity: sensor.stromzahler_power_consumption
    color_circle: true
  solar:
    entity: input_number.testwatt
```

Here are the test results:
- Solar Power 0 and Grid 0 > Increment Solar power > dots star moving > reverting solar power to 0 > value of solar power stays and will only change again when > 0
- Solar Power 0 and Grid 10 > grid dots move > Increment solar power > solar dots star moving too > reverting solar power to 0 > solar dots stop moving and value changes to 0, grid dots keep moving
- Solar Power 10 and Grid 10 > bot dots move > change  grid to 0 > grid dot stops, solar dot continues > change grid back to 10 > grid dots move again

So from what I'm seeing here is, that you card can not handle a state where all values are 0 and also the solar indicator will only change to 0 if the grid value is > 0 itself. Sure this might be a state which rarely occurs, but the solar panels might report 0 and the meter sensor needs some time to go from 0 (from the state where the panels were delivering enough to make it stop) to the wattage it's now delivering from the grid.

**Comment by flixlix** [comment]
Ok, that narrows it down. Though I still find it weird that with the same configuration and version of the card, I'm not experiencing this.
Does this issue still occur when you use the new flow rate model?

**Comment by jzielke84** [comment]
> Does this issue still occur when you use the new flow rate model?
What's that? Please specify.

**Comment by flixlix** [comment]
https://github.com/flixlix/power-flow-card-plus#new-flow-formula

**Comment by jzielke84** [comment]
Tried it like this, still the same:
```yaml
type: custom:power-flow-card-plus
entities:
  grid:
    entity: sensor.stromzahler_power_consumption
    color_circle: true
  solar:
    entity: input_number.testwatt
kw_decimals: 1
watt_threshold: 999
clickable_entities: true
title: Stromverbrauch
calculate_flow_rate: tru
min_expected_power: 0
```

**Comment by flixlix** [comment]
Should be like this:

```yaml
type: custom:power-flow-card-plus
entities:
  grid:
    entity: sensor.stromzahler_power_consumption
    color_circle: true
  solar:
    entity: input_number.testwatt
kw_decimals: 1
watt_threshold: 999
clickable_entities: true
title: Stromverbrauch
min_expected_power: 0
use_new_flow_rate_model: true
```

**Comment by jzielke84** [comment]
Sorry, still need my second coffee ☕  ;-) 

Yay that works! Thank you so much for your help. Maybe this model calculation should be made the default (e.g. true)?

**Comment by flixlix** [comment]
I'm glad that worked :)

Yeah, I thought about having the new flow rate model be default...

The issue is still not solved for the old flow rate model, so I'll try and see what the issue is there. For now, thank you very much for your cooperation 😊 and sorry for the thousands of questions from my side 😅

**Comment by jzielke84** [comment]
Never mind the questions, that's how you sometimes have to narrow down things. I bought you some "coffee". You deserve it! I'll leave this bug open for you to decide when/if it should be closed.

**Comment by flixlix** [comment]
Should be fixed in this PR #142 :)
